packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1704,43 +1705,76 @@ String nextPlatform(String currentPlatform, FeatureFlags featureFlags) {
   }
 }
 
+/// A manager for a devtools debugger instance.
+///
+/// NB: This class is overriden in google3.
 class DevtoolsLauncher {
-  io.HttpServer _devtoolsServer;
+  DevtoolsLauncher({
+    @required ProcessManager processManager,
+    @required String pubExecutable,
+    @required Logger logger,
+  }) : _processManager = processManager,
+       _pubExecutable = pubExecutable,
+       _logger = logger;
+
+  final ProcessManager _processManager;
+  final String _pubExecutable;
+  final Logger _logger;
+
+  io.Process _devtoolsProcess;
+  Uri _devtoolsUri;
+
+  static final RegExp _devtoolsPattern = RegExp(r'Serving DevTools at ((http|//)[a-zA-Z0-9:/=_\-\.\[\]]+)');
+
+  /// Launch a Dart DevTools process, optionally targeting a specific VM Service URI if
+  /// [observatoryAddress] is non-null.
   Future<void> launch(Uri observatoryAddress) async {
     try {
-      await serve();
-      await devtools_server.launchDevTools(
-        <String, dynamic>{
-          'reuseWindows': true,
-        },
-        observatoryAddress,
-        'http://${_devtoolsServer.address.host}:${_devtoolsServer.port}',
-        false,  // headless mode,
-        false,  // machine mode
-      );
+      _devtoolsProcess = await _processManager.start(<String>[
+        _pubExecutable,
+        'global',
+        'run',
+        'devtools',
+        '-b',
+        if (observatoryAddress != null)
+          '--vm-uri=$observatoryAddress',
+      ]);
+      final Completer<Uri> completer = Completer<Uri>();
+      _devtoolsProcess.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String line) {
+          final Match match = _devtoolsPattern.firstMatch(line);
+          if (match != null) {
+            completer.complete(Uri.parse(match[1]));
+          }
+          _logger.printTrace(line);
+        });
+      _devtoolsProcess.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(_logger.printError);
+      _devtoolsUri = await completer.future;
     } on Exception catch (e, st) {
-      globals.printTrace('Failed to launch DevTools: $e\n$st');
+      _logger.printError('Failed to launch DevTools: $e', stackTrace: st);
     }
   }
 
+  /// Serve Dart DevTools and return the host and port they are available on.
   Future<DevToolsServerAddress> serve() async {
-    try {
-      _devtoolsServer ??= await devtools_server.serveDevTools(
-        enableStdinCommands: false,
-      );
-      return DevToolsServerAddress(_devtoolsServer.address.host, _devtoolsServer.port);
-    } on Exception catch (e, st) {
-      globals.printTrace('Failed to serve DevTools: $e\n$st');
+    await launch(null);
+    if (_devtoolsUri == null) {
       return null;
     }
+    return DevToolsServerAddress(_devtoolsUri.host, _devtoolsUri.port);
   }
 
   Future<void> close() async {
-    await _devtoolsServer?.close();
-    _devtoolsServer = null;
+    _devtoolsProcess.kill();
```

Comment:
This does not seem to kill the devtools server, i would be worried about leaking it. FYI @jacob314

Reply:
we can send a command to the devtools server on stdout asking it to exit. That would be a couple line change to devtools.